File: app/src/main/java/tw/firemaples/onscreenocr/floatingviews/screencrop/DialogView.java
package tw.firemaples.onscreenocr.floatingviews.screencrop;

import android.content.Context;
import android.view.View;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.TextView;

import tw.firemaples.onscreenocr.R;
import tw.firemaples.onscreenocr.floatingviews.FloatingView;

/**
 * Created by firemaples on 22/11/2016.
 */

public class DialogView extends FloatingView {
    private TextView tv_dialogTitle, tv_dialogContent;
    private Button bt_dialogOk, bt_dialogCancel;

    private OnDialogViewCallback callback = new OnDialogViewCallback();

    public DialogView(Context context) {
        super(context);
        setViews(getRootView());
    }

    @Override
    protected int getLayoutId() {
        return R.layout.view_dialog;
    }

    @Override
    protected int getLayoutSize() {
        return WindowManager.LayoutParams.MATCH_PARENT;
    }

    private void setViews(View rootView) {
        tv_dialogTitle = (TextView) rootView.findViewById(R.id.tv_dialogTitle);
        tv_dialogContent = (TextView) rootView.findViewById(R.id.tv_dialogContent);

        bt_dialogOk = (Button) rootView.findViewById(R.id.bt_dialogOk);
        bt_dialogCancel = (Button) rootView.findViewById(R.id.bt_dialogCancel);

        bt_dialogOk.setOnClickListener(onClickListener);
        bt_dialogCancel.setOnClickListener(onClickListener);
    }

    public void setType(Type type) {
        switch (type) {
            case CONFIRM_ONLY:
                bt_dialogOk.setVisibility(View.VISIBLE);
                bt_dialogCancel.setVisibility(View.GONE);
                break;
            case CANCEL_ONLY:
                bt_dialogOk.setVisibility(View.GONE);
                bt_dialogCancel.setVisibility(View.VISIBLE);
                break;
            case CONFIRM_CANCEL:
                bt_dialogOk.setVisibility(View.VISIBLE);
                bt_dialogCancel.setVisibility(View.VISIBLE);
                break;
        }
    }

    public void reset() {
        callback = new OnDialogViewCallback();
        bt_dialogOk.setText(android.R.string.ok);
        bt_dialogCancel.setText(android.R.string.cancel);
    }

    @Override
    public boolean onBackButtonPressed() {
        callback.onCancelClicked(this);
        return true;
    }

    public void setCallback(OnDialogViewCallback callback) {
        this.callback = callback;
    }

    public void setTitle(String title) {
        tv_dialogTitle.setText(title);
    }

    public void setContentMsg(String msg) {
        tv_dialogContent.setText(msg);
    }

    public Button getOkBtn() {
        return bt_dialogOk;
    }

    public Button getCancelBtn() {
        return bt_dialogCancel;
    }

    private View.OnClickListener onClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            int id = v.getId();
            if (id == R.id.bt_dialogOk) {
                if (callback != null) {
                    callback.onConfirmClick(DialogView.this);
                } else {
                    detachFromWindow();
                }
            } else if (id == R.id.bt_dialogCancel) {
                if (callback != null) {
                    callback.onCancelClicked(DialogView.this);
                } else {
                    detachFromWindow();
                }
            }
        }
    };

    public enum Type {
        CONFIRM_ONLY, CANCEL_ONLY, CONFIRM_CANCEL
    }

    public static class OnDialogViewCallback {
        public void onConfirmClick(DialogView dialogView) {
            dialogView.detachFromWindow();
        }

        public void onCancelClicked(DialogView dialogView) {
            dialogView.detachFromWindow();
        }
    }
}

File: main/src/main/java/tw/firemaples/onscreenocr/floatings/dialog/DialogView.kt
package tw.firemaples.onscreenocr.floatings.dialog

import android.content.Context
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.FrameLayout
import android.widget.TextView
import kotlinx.coroutines.suspendCancellableCoroutine
import tw.firemaples.onscreenocr.R
import tw.firemaples.onscreenocr.floatings.base.FloatingView
import tw.firemaples.onscreenocr.utils.clickOnce
import tw.firemaples.onscreenocr.utils.setTextOrGone
import tw.firemaples.onscreenocr.utils.showOrHide
import kotlin.coroutines.resume

open class DialogView(context: Context) :
    FloatingView(context) {
    override val layoutId: Int
        get() = R.layout.floating_dialog
    override val layoutWidth: Int
        get() = WindowManager.LayoutParams.MATCH_PARENT
    override val layoutHeight: Int
        get() = WindowManager.LayoutParams.MATCH_PARENT

    private val viewRoot: View = rootView.findViewById(R.id.viewRoot)
    private val tvTitle: TextView = rootView.findViewById(R.id.tv_title)
    private val tvMessage: TextView = rootView.findViewById(R.id.tv_message)
    private val viewContentViewWrapper: FrameLayout =
        rootView.findViewById(R.id.view_contentViewWrapper)
    private val btOk: Button = rootView.findViewById(R.id.bt_ok)
    private val btCancel: Button = rootView.findViewById(R.id.bt_cancel)

    var onButtonOkClicked: (() -> Unit)? = null
    var onButtonCancelClicked: (() -> Unit)? = null

    init {
        setViews()
    }

    private fun setViews() {
        btOk.clickOnce {
            detachFromScreen()
            onButtonOkClicked?.invoke()
        }
        btCancel.clickOnce {
            detachFromScreen()
            onButtonCancelClicked?.invoke()
        }
    }

    fun setTitle(title: String?) {
        tvTitle.setTextOrGone(title)
    }

    fun setMessage(message: String) {
        tvMessage.setTextOrGone(message)
    }

    fun setButtonOkText(text: String) {
        btOk.text = text
    }

    fun setButtonCancelText(text: String) {
        btCancel.text = text
    }

    fun setContentView(
        view: View,
        layoutParams: FrameLayout.LayoutParams = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        )
    ) {
        tvMessage.visibility = View.GONE
        viewContentViewWrapper.removeAllViews()
        viewContentViewWrapper.addView(view, layoutParams)
        viewContentViewWrapper.visibility = View.VISIBLE
    }

    fun setDialogType(dialogType: DialogType) {
        btOk.showOrHide(dialogType == DialogType.CONFIRM_ONLY || dialogType == DialogType.CONFIRM_CANCEL)
        btCancel.showOrHide(dialogType == DialogType.CANCEL_ONLY || dialogType == DialogType.CONFIRM_CANCEL)
    }

    fun setCancelByClickingOutside(cancelByClickingOutside: Boolean) {
        if (cancelByClickingOutside) {
            rootView.clickOnce {
                detachFromScreen()
                onButtonCancelClicked?.invoke()
            }
        } else {
            rootView.clickOnce {}
        }
    }

    enum class DialogType {
        CONFIRM_ONLY,
        CANCEL_ONLY,
        CONFIRM_CANCEL,
    }
}

suspend fun Context.showDialog(
    title: String? = null,
    message: String,
    dialogType: DialogView.DialogType,
    cancelByClickingOutside: Boolean = false,
): Boolean = suspendCancellableCoroutine { cont ->
    DialogView(this).apply {
        setTitle(title)
        setMessage(message)
        setDialogType(dialogType)
        setCancelByClickingOutside(cancelByClickingOutside)

        onButtonOkClicked = {
            cont.resume(true)
        }

        onButtonCancelClicked = {
            cont.resume(false)
        }

        cont.invokeOnCancellation { detachFromScreen() }

        attachToScreen()
    }
}

suspend fun Context.showErrorDialog(error: String): Boolean =
    suspendCancellableCoroutine { cont ->
        DialogView(this).apply {
            setTitle(getString(R.string.title_error))
            setMessage(error)
            setDialogType(DialogView.DialogType.CONFIRM_ONLY)
            setCancelByClickingOutside(false)

            onButtonOkClicked = { cont.resume(false) }

            cont.invokeOnCancellation { detachFromScreen() }
        }.attachToScreen()
    }
